### PR TITLE
feat: Created hedera-addressbook-service package for Dynamic Address Book P2.

### LINKED ISSUE #13150: Add hedera-addressbook-service package
Created hedera-addressbook-service package for Dynamic Address Book P2.

### PR DESCRIPTION
**Pull request by iwsimon**
**Description**:

**Related issue(s)**:

Fixes #13150

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by tomzhenghedera — DISMISSED** [review]

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ServiceLoader;
+import java.util.Set;
+
+/**
+ * Implements the HAPI <a
+ * href="https://github.com/hashgraph/hedera-protobufs/blob/main/services/address_book_service.proto">Consensus
+ * Service</a>.
+ */
+public interface AddressbookService extends Service {
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java:31]:
```suggestion
public interface AddressBookService extends Service {
```

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    String NAME = "AddressbookService";
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java:33]:
```suggestion
    String NAME = "AddressBookService";
```

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookServiceDefinition.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import com.hedera.pbj.runtime.RpcMethodDefinition;
+import com.hedera.pbj.runtime.RpcServiceDefinition;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Set;
+
+/**
+ * The Addressbook Service provides the ability for Hedera Hashgraph to provide facilitate changes to the nodes used across the Hedera network.
+ */
+@SuppressWarnings("java:S6548")
+public final class AddressbookServiceDefinition implements RpcServiceDefinition {
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookServiceDefinition.java:32]:
```suggestion
public final class AddressBookServiceDefinition implements RpcServiceDefinition {
```

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import com.hedera.node.app.spi.ServiceFactory;
+import com.hedera.pbj.runtime.RpcServiceDefinition;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ServiceLoader;
+import java.util.Set;
+
+/**
+ * Implements the HAPI <a
+ * href="https://github.com/hashgraph/hedera-protobufs/blob/main/services/address_book_service.proto">Consensus
+ * Service</a>.
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java:29]:
```suggestion
 * href="https://github.com/hashgraph/hedera-protobufs/blob/main/services/address_book_service.proto"> Address Book Service</a>.
```

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/ReadableNodeStore.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    /**
+     * Returns the node needed. If the node doesn't exist returns failureReason. If the
+     * node exists , the failure reason will be null.
+     *
+     * @param nodeId node id being looked up
+     * @return node's metadata
+     */
+    @Nullable
+    Node getNode(final long nodeId);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/ReadableNodeStore.java:38]: Could probably just make this method `get(final long nodeId)` to follow same pattern in other services

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/ReadableNodeStore.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    /**
+     * Returns the number of nodes in the state.
+     * @return the number of nodes in the state.
+     */
+    long sizeOfState();
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/ReadableNodeStore.java:44]: We need `warm()` method if this is a virtual map

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java (hunk 1 above)
Comment (iwsimon) [reply hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java]: Updated.

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java (hunk 2 above)
Comment (iwsimon) [reply hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java]: Updated.

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java (hunk 4 above)
Comment (iwsimon) [reply hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressbookService.java]: Updated.

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/ReadableNodeStore.java (hunk 6 above)
Comment (iwsimon) [reply hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/ReadableNodeStore.java:44]: Added.

File: hedera-node/hedera-addressbook-service/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+plugins { id("com.hedera.hashgraph.conventions") }
+
+description = "Hedera Addressbook Service API"
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service/build.gradle.kts:19]:
```suggestion
description = "Hedera AddressBook Service API"
```

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2023-2024 Hedera Hashgraph, LLC
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java:2]: Shouldn't these copyrights be 2024-2025 in all new files?

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM! Just have two comments. Thanks @iwsimon

File: hedera-node/hedera-addressbook-service/build.gradle.kts (hunk 7 above)
Comment (iwsimon) [reply hedera-node/hedera-addressbook-service/build.gradle.kts]: Updated.

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java (hunk 8 above)
Comment (iwsimon) [reply hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java]: Changed.

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM ! Thanks @iwsimon

**Review by Neeharika-Sompalli — APPROVED** [review]

**Review by povolev15 — APPROVED** [review]
LGTM TY @iwsimon

**Review by JeffreyDallas — APPROVED** [review]

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import com.hedera.hapi.node.transaction.Query;
+import com.hedera.hapi.node.transaction.Response;
+import com.hedera.hapi.node.transaction.TransactionResponse;
+import com.hedera.pbj.runtime.RpcMethodDefinition;
+import com.hedera.pbj.runtime.RpcServiceDefinition;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Set;
+
+/**
+ * The Addressbook Service provides the ability for Hedera Hashgraph to provide facilitate changes to the nodes used across the Hedera network.
```
Comment (kimbor) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java:29]:
```suggestion
 * The Addressbook Service provides the ability for Hedera Hashgraph to facilitate changes to the nodes used across the Hedera network.
```

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+public final class AddressBookServiceDefinition implements RpcServiceDefinition {
+    public static final AddressBookServiceDefinition INSTANCE = new AddressBookServiceDefinition();
+
+    private static final Set<RpcMethodDefinition<?, ?>> methods = Set.of(
+
+            // Prepare to add a new node to the network.
+            // When a valid council member initiates a HAPI transaction to add a new node,
+            // then the network should acknowledge the transaction and update the network’s Address Book within 24
+            // hours.
+            // The added node will not be active until the network is upgradede
```
Comment (kimbor) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java:41]:
```suggestion
            // The added node will not be active until the network is upgraded
```

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+            // Request is [NodeCreateTransactionBody](#proto.NodeCreateTransactionBody)
+            //
+            new RpcMethodDefinition<>("createNode", Transaction.class, TransactionResponse.class),
+            // Prepare to update the node to the network.
+            // The node will not be updated until the network is upgraded.
+            // Request is [NodeUpdateTransactionBody](#proto.NodeUpdateTransactionBody)
+            //
+            new RpcMethodDefinition<>("updateNode", Transaction.class, TransactionResponse.class),
+            // Prepare to delete the node to the network.
```
Comment (kimbor) [inline hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java:50]:
```suggestion
            // Prepare to delete the node from the network.
```

**Review by kimbor — APPROVED** [review]

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java (hunk 9 above)
Comment (iwsimon) [reply hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java:29]: I will update those in the next PR. Thanks @kimbor

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java (hunk 10 above)
Comment (iwsimon) [reply hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java:41]: Updated in next PR.

File: hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java (hunk 11 above)
Comment (iwsimon) [reply hedera-node/hedera-addressbook-service/src/main/java/com/hedera/node/app/service/addressbook/AddressBookServiceDefinition.java:50]: Updated in next PR.

**Pull request merged**